Build fails on Windows (current master)

I wanted to try the new design of hybrid and standard evaluation, but couldn't due to this error on my Windows; ``undefined reference to `std::tr1::hash<Rcomplex>::operator()(Rcomplex) const'``. I think you can see the full build log via AppVeyer, but here's some last lines in case this is useful:

``` 
c:/Rtools/mingw_64/bin/g++ -shared -s -static-libgcc -o dplyr.dll tmp.def RcppExports.o address.o api.o arrange.o between.o bind.o distinct.o encoding.o filter.o group_by.o group_indices.o init.o join.o join_exports.o mutate.o rlang-export.o rlang.o select.o set.o summarise.o test.o utils-bindings.o utils.o window.o -LC:/PROGRA~1/R/R-35~1.1/bin/x64 -lR
mutate.o:mutate.cpp:(.text$_ZNKSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE5countERKS1_[_ZNKSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE5countERKS1_]+0x27): undefined reference to `std::tr1::hash<Rcomplex>::operator()(Rcomplex) const'
mutate.o:mutate.cpp:(.text$_ZNSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE9_M_rehashEy[_ZNSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE9_M_rehashEy]+0x69): undefined reference to `std::tr1::hash<Rcomplex>::operator()(Rcomplex) const'
mutate.o:mutate.cpp:(.text$_ZNSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE9_M_insertERKS1_NS_17integral_constantIbLb1EEE[_ZNSt3tr110_HashtableI8RcomplexS1_SaIS1_ESt9_IdentityIS1_ESt8equal_toIS1_ENS_4hashIS1_EENS_8__detail18_Mod_range_hashingENS9_20_Default_ranged_hashENS9_20_Prime_rehash_policyELb0ELb1ELb1EE9_M_insertERKS1_NS_17integral_constantIbLb1EEE]+0x34): undefined reference to `std::tr1::hash<Rcomplex>::operator()(Rcomplex) const'
collect2.exe: error: ld returned 1 exit status
no DLL was created
ERROR: compilation failed for package 'dplyr'
```

I don't understand what's happening here, but, if I use `boost::unordered_map` and `boost::unordered_set` instead of `RCPP_UNORDERED_MAP` and `RCPP_UNORDERED_SET`, the build succeeded (the tests fail, though...). More precisely, removing these `if`s here:

https://github.com/tidyverse/dplyr/blob/53ed25b65a8db4b619ad7212863c6532b66c8021/inst/include/tools/hash.h#L7-L9

https://github.com/tidyverse/dplyr/blob/53ed25b65a8db4b619ad7212863c6532b66c8021/inst/include/tools/hash.h#L16-L18

`RCPP_UNORDERED_MAP` and `RCPP_UNORDERED_SET` were introduced in https://github.com/tidyverse/dplyr/commit/f82f55d8337f1bccc9f6f0e751639f42dfca62f8, which addresses https://github.com/tidyverse/dplyr/issues/114#issuecomment-28428103. Is this workaround still needed? Note that I'm using Rtools 3.5, so I don't know if this works on the older version of R on Windows.

--- a/inst/include/tools/hash.h
+++ b/inst/include/tools/hash.h
@@ -21,6 +21,37 @@
 #endif
 #endif // #ifndef dplyr_hash_set
 
+// FIXME: remove this when Rcpp provides a hash function for Rcomplex
+#if defined(_WIN32)
+#if __cplusplus >= 201103L
+namespace std {
+  template<>
+  struct hash<Rcomplex> {
+    std::size_t operator()(const Rcomplex& cx) const {
+      std::hash<double> hasher;
+      size_t seed = hasher(cx.r);
+      boost::hash_combine(seed, hasher(cx.i));
+      return seed;
+    }
+  };
+}
+#elif defined(HAS_TR1_UNORDERED_SET)
+namespace std {
+  namespace tr1 {
+    template<>
+    struct hash<Rcomplex> {
+      std::size_t operator()(const Rcomplex& cx) const {
+	std::tr1::hash<double> hasher;
+	size_t seed = hasher(cx.r);
+	boost::hash_combine(seed, hasher(cx.i));
+	return seed;
+      }
+    };
+  }
+}
+#endif
+#endif
+
 inline std::size_t hash_value(const Rcomplex& cx) {
   boost::hash<double> hasher;
   size_t seed = hasher(cx.r);

